### ISSUE TITLE
Phase 1: NER Entity Extraction Pipeline - Data Collection

### ISSUE THREAD

**Issue by holden** [razrfly/volfefe#43]
# Phase 1: Named Entity Recognition (NER) - Entity Extraction Pipeline

**Status**: Ready for Implementation  
**Priority**: Critical - Foundation for Asset Linking  
**Branch**: `10-27-asset_content`  
**Type**: Research & Data Collection

---

## 🎯 Mission

**Add a 4th model to our existing multi-model ML pipeline to extract companies, industries, and countries from content.**

This is PURE entity extraction - we're collecting data to understand WHAT entities appear in our content before we build Phase 2 (asset linking).

---

### What We Have ✅

- **Multi-Model Sentiment Pipeline**: 3 BERT models running in parallel (`classify_multi_model.py`)
  - DistilBERT (SST-2)
  - Twitter-RoBERTa  
  - FinBERT
- **9,000+ Assets**: Companies loaded with symbols, names, exchanges
- **ContentTargets Schema**: Ready to store content→asset relationships
- **Content Corpus**: Existing classified content ready for entity extraction

### What We're Building 🔨

**Add NER Model #4** to extract:
1. **Organizations/Companies**: Tesla, Apple, Goldman Sachs, Fed
2. **Locations**: United States, China, San Francisco, Europe
3. **Miscellaneous**: Could capture industries, products, events

### Why This Matters

**Before Phase 1:**
```
Content: "Tesla stock crashes 15% after Musk controversy"
Classification: NEGATIVE (0.92 confidence)
Entities: ??? UNKNOWN ???
```

**After Phase 1:**
```
Content: "Tesla stock crashes 15% after Musk controversy"  
Classification: NEGATIVE (0.92 confidence)
Entities: [
  {text: "Tesla", type: "ORG", confidence: 0.95},
  {text: "Musk", type: "PER", confidence: 0.89}
]
```

---

### Recommended Model: `dslim/bert-base-NER`

**Why This Model:**
- ✅ **State-of-the-art CoNLL-2003 performance** (94-96% F1)
- ✅ **Four entity types**: PER, ORG, LOC, MISC
- ✅ **Fast inference** (~50-100ms per document)
- ✅ **Production-ready** (10M+ downloads on Hugging Face)
- ✅ **No special configuration** required
- ✅ **Perfect fit** for our existing pipeline architecture

**Entity Classes:**
- `ORG` - Organizations, companies, agencies (Tesla, Apple, Federal Reserve)
- `LOC` - Locations, countries, cities, regions (United States, San Francisco, Europe)
- `PER` - Person names (Elon Musk, Jerome Powell)
- `MISC` - Miscellaneous entities (products, events, might capture industries)

**Training Data:** CoNLL-2003 (English news articles)

**Performance Metrics:**
- Precision: 0.9507
- Recall: 0.9533
- F1: 0.9520
- Accuracy: 0.9868

### Alternative Models Considered

| Model | Pros | Cons | Recommendation |
|-------|------|------|----------------|
| `xlm-roberta-large-finetuned-conll03-english` | Multilingual, high accuracy | Larger/slower (1.1GB vs 420MB) | Phase 2+ if needed |
| `Jean-Baptiste/roberta-large-ner-english` | High performance | Larger model | Phase 2+ if needed |
| `dslim/distilbert-NER` | Faster inference | Slightly lower accuracy | Phase 2+ for speed optimization |

---

### Step 1: Add NER Model to Python Pipeline (2-3 hours)

**Goal**: Extend `priv/ml/classify_multi_model.py` to include NER extraction

**Approach**: Treat NER like a 4th sentiment model

**Tasks**:
- [ ] Add `dslim/bert-base-NER` to model loading
- [ ] Create `extract_entities(text)` function using `pipeline("ner")`
- [ ] Handle entity aggregation (subword tokens → full entity spans)
- [ ] Return structured entity data with confidence scores

**Output Format**:
```json
{
  "results": [...existing sentiment models...],
  "entities": {
    "model_id": "bert_base_ner",
    "model_version": "dslim/bert-base-NER",
    "extracted": [
      {
        "text": "Tesla",
        "type": "ORG",
        "confidence": 0.9531,
        "start": 0,
        "end": 5,
        "context": "Tesla stock crashes 15%"
      },
      {
        "text": "United States",
        "type": "LOC", 
        "confidence": 0.9872,
        "start": 45,
        "end": 58,
        "context": "...concerns in United States markets"
      }
    ],
    "stats": {
      "total_entities": 2,
      "by_type": {"ORG": 1, "LOC": 1, "PER": 0, "MISC": 0}
    },
    "meta": {
      "processing": {
        "latency_ms": 87,
        "timestamp": "2025-10-27T12:34:56Z"
      }
    }
  },
  "total_latency_ms": 2387
}
```

**Implementation Notes**:
- Use `aggregation_strategy="simple"` to merge subword tokens
- Extract context window (±20 chars) around each entity
- Capture ALL metadata (confidence, positions, types)
- Handle multi-word entities correctly
- No filtering by confidence yet - store everything

### Step 2: Elixir Integration (1-2 hours)

**Goal**: Call extended Python script from Elixir, parse entity results

**Module**: `VolfefeMachine.Intelligence.EntityExtractor`

**Tasks**:
- [ ] Extend `MultiModelClient` to parse `entities` field
- [ ] Create Ecto schema for temporary entity storage (optional)
- [ ] Add `EntityExtractor.extract(content_id)` wrapper function
- [ ] Handle errors gracefully (model loading failures, parsing errors)

**API**:
```elixir
# Extract entities from content
{:ok, result} = Intelligence.EntityExtractor.extract(content_id)

result.entities
# => [
#   %{text: "Tesla", type: "ORG", confidence: 0.95, ...},
#   %{text: "United States", type: "LOC", confidence: 0.98, ...}
# ]

result.stats
# => %{total: 2, by_type: %{"ORG" => 1, "LOC" => 1}}
```

### Step 3: Data Collection Script (1 hour)

**Goal**: Run NER across ALL existing content to collect entity distribution data

**Mix Task**: `mix volfefe.extract_entities`

**Tasks**:
- [ ] Fetch all classified content from database
- [ ] Run entity extraction on each
- [ ] Store results in JSONB for analysis (OR just collect stats)
- [ ] Generate comprehensive report

**Output**:
```
=====================================
 Entity Extraction Report
=====================================

Processed: 1,247 content items
Total Entities: 4,892
Extraction Time: 3m 42s
Avg per content: 3.9 entities

--- Entity Type Distribution ---
ORG:  2,156 (44.1%)
LOC:  1,634 (33.4%)
PER:    892 (18.2%)
MISC:   210 ( 4.3%)

--- Top Organizations Extracted ---
1. Tesla         - 234 mentions
2. Apple         - 189 mentions
3. Microsoft     - 156 mentions
4. Amazon        - 143 mentions
5. Google        - 128 mentions
...top 20...

--- Top Locations Extracted ---
1. United States - 456 mentions
2. China         - 234 mentions
3. Europe        - 145 mentions
4. California    - 98 mentions
5. New York      - 87 mentions
...top 20...

--- Confidence Distribution ---
High (>0.9):     3,892 (79.6%)
Medium (0.7-0.9): 876 (17.9%)
Low (<0.7):       124 ( 2.5%)

--- Extraction Quality ---
Avg confidence: 0.9134
Latency avg:    89ms per content
Total processed: 100%
Failed:         0 (0%)
```

### Step 4: Analysis & Validation (2-3 hours)

**Goal**: Understand what we're extracting before building asset linking

**Tasks**:
- [ ] **Manual Review**: Inspect 50 random entity extractions
- [ ] **Accuracy Assessment**: How many ORG entities are actual companies vs other orgs?
- [ ] **Coverage Analysis**: Do we extract major companies mentioned in content?
- [ ] **Data Quality**: Are confidence scores reliable indicators?
- [ ] **Entity Patterns**: What types of ORGs appear most? (companies, governments, agencies)
- [ ] **Matching Feasibility**: How many extracted ORGs match our 9K assets?

**Deliverable**: Written analysis answering:
1. What percentage of ORG entities are tradable companies?
2. What percentage are other orgs (Fed, SEC, governments)?
3. Are LOC entities useful for market analysis?
4. Should we filter by confidence? What threshold?
5. What challenges will Phase 2 (asset linking) face?

---

## 🎯 Success Criteria

**Phase 1 Complete When:**

1. ✅ NER model integrated into `classify_multi_model.py`
2. ✅ `EntityExtractor.extract/1` works in Elixir
3. ✅ Mix task processes all existing content
4. ✅ Comprehensive entity extraction report generated
5. ✅ Manual validation of 50+ samples completed
6. ✅ Analysis document written with Phase 2 recommendations

**Performance Targets:**
- Extract entities in <100ms per content (NER model only)
- Total pipeline <2.5 seconds (sentiment + NER)
- Process 1,000 content items in <45 minutes
- Entity extraction success rate >95%

**Quality Targets:**
- NER confidence >0.85 average
- Manual validation accuracy >85%
- Clear understanding of entity distribution
- Documented challenges for Phase 2

---

## 🚫 Non-Goals (Explicitly NOT Phase 1)

**We are NOT building:**
- ❌ Asset linking/matching logic
- ❌ ContentTargets creation
- ❌ Fuzzy name matching
- ❌ Entity disambiguation
- ❌ Industry/sector classification  
- ❌ Alias tables
- ❌ Query APIs

**Phase 2 will handle linking extracted entities to assets.**

---

### Why Not Use spaCy?

- ❌ Different dependency stack (adds complexity)
- ❌ Different pipeline architecture
- ❌ Would require separate Python environment management
- ✅ Transformers already proven in our stack
- ✅ BERT-based NER more consistent with sentiment models

### Why Not Multiple NER Models?

- ❌ Overkill for Phase 1 (just collecting data)
- ❌ Consensus logic adds complexity
- ✅ Single high-quality model sufficient for analysis
- ✅ Can add more models in Phase 2 if needed

### Why Not Store Entities in Database?

**Decision**: Generate report, analyze, THEN decide storage strategy

**Rationale**:
- ✅ Faster iteration (no schema design)
- ✅ Flexibility to change approach
- ✅ Can add storage in Phase 2 after validation
- ❌ Can't query entities easily (acceptable for Phase 1)

**Alternative**: Store in `meta` JSONB field on Content table
- Could add `entities` JSONB column to `contents`
- Queryable but not relational
- Good middle ground for Phase 1

---

### Input Content
```
"Tesla stock tumbled 12% today as Elon Musk's controversial tweet 
sparked concerns about the company's future. Analysts in the United States 
and Europe are worried about automotive sector stability."
```

### After Sentiment Analysis (Existing)
```elixir
%{
  results: [
    %{model_id: "distilbert", sentiment: "negative", confidence: 0.9812},
    %{model_id: "twitter_roberta", sentiment: "negative", confidence: 0.9654},
    %{model_id: "finbert", sentiment: "negative", confidence: 0.9201}
  ],
  consensus: %{sentiment: "negative", confidence: 0.9556}
}
```

### After NER Extraction (NEW - Phase 1)
```elixir
%{
  entities: [
    %{text: "Tesla", type: "ORG", confidence: 0.9531, start: 0, end: 5},
    %{text: "Elon Musk", type: "PER", confidence: 0.9802, start: 34, end: 43},
    %{text: "United States", type: "LOC", confidence: 0.9912, start: 115, end: 128},
    %{text: "Europe", type: "LOC", confidence: 0.9845, start: 133, end: 139},
    %{text: "automotive sector", type: "MISC", confidence: 0.8567, start: 162, end: 179}
  ],
  stats: %{
    total_entities: 5,
    by_type: %{"ORG" => 1, "LOC" => 2, "PER" => 1, "MISC" => 1}
  }
}
```

### Phase 1 Analysis Output
```
Entities Extracted: 5
Companies Detected: 1 (Tesla)
Locations: 2 (United States, Europe)
Industry Keywords: 1 (automotive sector)

PHASE 2 PREVIEW (not implemented):
  - "Tesla" → Match to Asset{id: 123, symbol: "TSLA"}
  - Create ContentTarget{content_id: 1, asset_id: 123, confidence: 0.95}
```

---

## 🔬 Research Questions Phase 1 Answers

By the end of Phase 1, we'll know:

1. **Entity Distribution**
   - How many companies vs other ORGs?
   - How many locations are countries vs cities?
   - Are MISC entities useful for industry classification?

2. **Extraction Quality**
   - Is 0.85 confidence threshold appropriate?
   - How accurate is the NER model on financial content?
   - What types of entities are commonly missed?

3. **Asset Linking Feasibility**
   - What percentage of extracted companies are in our asset database?
   - How many variations of company names do we see? (Tesla vs Tesla Inc)
   - Will fuzzy matching be sufficient or do we need alias tables?

4. **Phase 2 Scope**
   - Should we focus only on ORG → Asset linking?
   - Is LOC → Country → Market mapping valuable?
   - Should MISC → Industry classification be Phase 3?

---

## 🗓️ Timeline

**Total Estimated Time: 6-9 hours**

- **Step 1** (NER Integration): 2-3 hours
- **Step 2** (Elixir API): 1-2 hours  
- **Step 3** (Data Collection): 1 hour
- **Step 4** (Analysis): 2-3 hours

**Deliverables Timeline:**
- Day 1: Steps 1-2 (NER working end-to-end)
- Day 2: Step 3 (Data collection complete)
- Day 3: Step 4 (Analysis and Phase 2 planning)

---

### Model Documentation
- **dslim/bert-base-NER**: https://huggingface.co/dslim/bert-base-NER
- **Transformers NER Guide**: https://huggingface.co/docs/transformers/tasks/token_classification
- **Pipeline API**: https://huggingface.co/docs/transformers/main_classes/pipelines

### Technical References
- **CoNLL-2003 Dataset**: Standard NER benchmark
- **Entity Aggregation**: Handling subword tokenization
- **Token Classification**: Transformers task type

### Alternative Models (Phase 2+)
- `xlm-roberta-large-finetuned-conll03-english` - Multilingual
- `Jean-Baptiste/roberta-large-ner-english` - Higher accuracy
- `dslim/distilbert-NER` - Faster inference

---

## 🔗 Related Issues

- Issue #26: Assets table foundation (COMPLETED)
- Issue #34: Assets loaded (COMPLETED)
- Issue #36: Original NER vision (REFERENCE)
- Issue #42: Phase 2 - Asset linking pipeline (BLOCKED by this)

---

## 🎓 Learning Outcomes

By completing Phase 1, we'll:

1. Understand NER model capabilities and limitations
2. Know our entity distribution and patterns
3. Have data to make informed Phase 2 decisions
4. Validate feasibility of asset linking approach
5. Identify challenges early (aliases, ambiguity, coverage)

**Phase 2 will be MUCH easier with this foundation.**

---

## Labels

`enhancement` `research` `machine-learning` `intelligence` `critical` `phase-1`

**Comment by holden** [comment]
✅ Completed - NER pipeline operational with ner_worker.ex processing entities and content_targets schema storing extraction results.
